finality-aleph/src/testing/network.rs

```diff
@@ -0,0 +1,315 @@
+use futures::channel::{mpsc, oneshot};
+use sc_network::{Event, Multiaddr as ScMultiaddr, ObservedRole};
+use sc_service::TaskManager;
+use std::{
+    collections::{HashMap, HashSet},
+    time::Duration,
+};
+use tokio::{runtime::Handle, task::JoinHandle, time::timeout};
+
+const DEFAULT_TIMEOUT: Duration = Duration::from_secs(10);
+const NODES_N: usize = 3;
+type MockData = Vec<u8>;
+
+#[derive(Clone)]
+struct Authority {
+    pen: AuthorityPen,
+    addresses: Vec<ScMultiaddr>,
+    peer_id: PeerId,
+}
+
+impl Authority {
+    fn pen(&self) -> AuthorityPen {
+        self.pen.clone()
+    }
+
+    fn addresses(&self) -> Vec<ScMultiaddr> {
+        self.addresses.clone()
+    }
+
+    fn peer_id(&self) -> PeerId {
+        self.peer_id
+    }
+}
+
+impl NetworkIdentity for Authority {
+    fn identity(&self) -> (Vec<ScMultiaddr>, PeerId) {
+        (self.addresses.clone(), self.peer_id)
+    }
+}
+
+struct TestData {
+    pub authorities: Vec<Authority>,
+    pub authority_verifier: AuthorityVerifier,
+    pub session_manager: SessionManager<MockData>,
+    pub network: MockNetwork<NetworkData<MockData>>,
+    network_manager_exit_tx: oneshot::Sender<()>,
+    network_service_exit_tx: oneshot::Sender<()>,
+    network_manager_handle: JoinHandle<()>,
+    network_service_handle: JoinHandle<()>,
+    _task_manager: TaskManager,
+}
+
+async fn prepare_one_session_test_data() -> TestData {
+    let task_manager = TaskManager::new(Handle::current(), None).unwrap();
+    let (authority_pens, authority_verifier) = crypto_basics(NODES_N).await;
+    let authorities: Vec<_> = authority_pens
+        .into_iter()
+        .map(|(_, p)| {
+            let identity = MockNetworkIdentity::new().identity();
+            Authority {
+                pen: p,
+                addresses: identity.0,
+                peer_id: identity.1,
+            }
+        })
+        .collect();
+
+    // Prepare Network
+    let (event_stream_tx, event_stream_rx) = oneshot::channel();
+    let (network_manager_exit_tx, network_manager_exit_rx) = oneshot::channel();
+    let (network_service_exit_tx, network_service_exit_rx) = oneshot::channel();
+    let network = MockNetwork::new(event_stream_tx);
+
+    let (commands_for_network, commands_from_io) = mpsc::unbounded();
+    let (messages_for_network, messages_from_user) = mpsc::unbounded();
+    let (commands_for_service, commands_from_user) = mpsc::unbounded();
+    let (messages_for_service, commands_from_manager) = mpsc::unbounded();
+    let (messages_for_user, messages_from_network) = mpsc::unbounded();
+
+    let connection_io = ConnectionIO::new(
+        commands_for_network,
+        messages_for_network,
+        commands_from_user,
+        commands_from_manager,
+        messages_from_network,
+    );
+    let connection_manager = ConnectionManager::<Authority, MockData>::new(
+        authorities[0].clone(),
+        ConnectionManagerConfig::with_session_period(&SessionPeriod(10), &MillisecsPerBlock(1000)),
```

Comment:
Maybe you would like to move out those values as constants

Reply:
Done